### PR TITLE
Add 479 experimental module in bayesian problem

### LINKED ISSUE #479: Update the BayesianProblem to use the new sampler
Description:
-------------
Update the BayesainProblem class to have 2 behaviors. A default behavior to use the new samplers and a legacy behavior to use the old samplers (to ensure tests pass for now)

DoD:
------
- [ ] The BayesainProblem is updated to have the behavior described above
- [ ] Any related code is also updated (demos/tests)
- [ ] The updates are reviewed and merged

Note:
------
- (written by @amal-ghamdi) notes from @chaozg in https://github.com/CUQI-DTU/CUQI-Book/issues/40 could be relevant
- (written by @amal-ghamdi) Suggestion by @jakobsj, Since the mcmc module is experimental, we can make the default behaviour of BayesainProblem for now to be the old behavior and pass a flag if we want the new behavior (@nabriis)

### PR DESCRIPTION
**Pull request by nabriis**
Closes #479 

Best reviewed in split view.

Suggest to review by trying the new experimental mode in the CUQI book.

### CODE REVIEW

**Review by amal-ghamdi — CHANGES_REQUESTED** [review]
Many thanks @nabriis for this update!! It looks great, I tested it on training notebooks 3 and 4 and the results are below (results are consistent). There is the issue that the output of the sampling progress bar is across so many lines in the experimental case but was good in the legacy case. But I think this is out of the scope of this PR.

I also added few minor comments in the review.

Training Notebook B03 (from the book)
experimental = False (output of sampling progress is nice)

image

experimental = True  (output of sampling progress is very long)

image

Training Notebook B04 (from the book)

experimental =False
image
image

experimental = True

image

image

File: cuqi/problem/_problem.py [hunk 1]
```diff
@@ -719,31 +837,61 @@ def _check_posterior(posterior, prior_type=None, likelihood_type=None, model_typ
...
+            samples = sampler.sample(Ns, Nb)
+
+            # Print timing
+            print('Elapsed time:', time.time() - ti)
 
         return samples
 
-    def _determine_sampling_strategy(self):
+
+    def _determine_sampling_strategy(self, experimental=False):
```
[the remark this answers is not in the document]
Comment (nabriis) [reply cuqi/problem/_problem.py:894]: Yes that sounds good. Perhaps best once we have a clear indication the HybridGibbs is working well with NUTS/MH etc. And likely after #503 is at least merged.

File: cuqi/problem/_problem.py [hunk 2]
```diff
@@ -475,19 +485,37 @@ def _alternative_plot_UQ_for_variable(
...
+    def _sampleLinearRTO(self, Ns, Nb, callback=None, experimental=False):
 
-        # Start timing
-        ti = time.time()
+        if experimental:
 
-        # Sample
-        sampler = cuqi.sampler.LinearRTO(self.posterior, callback=callback)
-        samples = sampler.sample(Ns, Nb)
+            print("Using experimental LinearRTO sampler.")
```
[the remark this answers is not in the document]
Comment (nabriis) [reply cuqi/problem/_problem.py]: Done. Good suggestion.

File: cuqi/problem/_problem.py [hunk 3]
```diff
@@ -319,39 +322,43 @@ def sample_posterior(self, Ns, Nb=None, callback=None) -> cuqi.samples.Samples:
         print("!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!")
```
[the remark this answers is not in the document]
Comment (nabriis) [reply cuqi/problem/_problem.py]: Agreed. Done

File: tests/test_bayesian_inversion.py [hunk 4]
```diff
@@ -44,18 +45,27 @@ def test_TP_BayesianProblem_sample(copy_reference, TP_type, phantom, prior, Ns,
...
+    #if isinstance(prior, RegularizedGMRF): #Put the case you want to update for here.
     #    np.savez(ref_fname, median=med_xpos, sigma=sigma_xpos, lo95=lo95, up95=up95) #uncomment to update
+    
     ref_file = copy_reference(f"data/{ref_fname}.npz")
     ref = np.load(ref_file)
 
     # Check results with reference data
+    #if isinstance(prior, RegularizedGMRF) and not sys.platform.startswith('win') and experimental:
+    #    pass # Skip this case
+    #else:
```
[the remark this answers is not in the document]
Comment (nabriis) [reply tests/test_bayesian_inversion.py]: Removed. Fixed the issue but forgot to remove the commented code.

**Comment by nabriis** [comment]
Hi @amal-ghamdi  and @chaozg. Ready for review again.

**Review by chaozg — APPROVED** [review]
Thanks @nabriis . This PR looks very nice and I don't have any suggestions except for the following tiny comments:

- I'm curious if the statement here, "[Most functionality is currently only implemented for this simple case.](https://github.com/CUQI-DTU/CUQIpy/blob/cd99155177420d1f6751bd6cb5c7771718aa6336/cuqi/problem/_problem.py#L130)", is still true with Gibbs, particulary the new HybridGibbs?
- A tiny reminder for the future, we still use the `x0` keyword in `BayesianProblem`, instead of `initial_point`, and this is of course out of the scope of this PR.

File: cuqi/problem/_problem.py [hunk 5]
```diff
@@ -218,10 +218,10 @@ def ML(self, disp=True, x0=None) -> CUQIarray:
         """
         if disp:
             # Print warning to user about the automatic solver selection
-            print("!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!")
-            print("!!! Automatic solver selection is experimental. !!!")
-            print("!!!    Always validate the computed results.    !!!")
-            print("!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!")
+            print("!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!")
+            print("!!! Automatic solver selection is work-in-progress !!!")
```
Comment (amal-ghamdi) [inline cuqi/problem/_problem.py:222]:
```suggestion
            print("!!! Automatic solver selection is a work-in-progress !!!")
```

File: cuqi/problem/_problem.py [hunk 6]
```diff
@@ -254,10 +254,10 @@ def MAP(self, disp=True, x0=None) -> CUQIarray:
 
         if disp:
             # Print warning to user about the automatic solver selection
-            print("!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!")
-            print("!!! Automatic solver selection is experimental. !!!")
-            print("!!!    Always validate the computed results.    !!!")
-            print("!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!")
+            print("!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!")
+            print("!!! Automatic solver selection is work-in-progress !!!")
```
Comment (amal-ghamdi) [inline cuqi/problem/_problem.py:258]:
```suggestion
            print("!!! Automatic solver selection is a work-in-progress !!!")
```

File: cuqi/problem/_problem.py [hunk 7]
```diff
@@ -316,14 +316,16 @@ def sample_posterior(self, Ns, Nb=None, callback=None, experimental=False) -> cu
         """
 
         # Print warning to user about the automatic sampler selection
-        print("!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!")
-        print("!!! Automatic sampler selection is experimental. !!!")
-        print("!!!    Always validate the computed results.     !!!")
-        print("!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!")
+        print("!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!")
+        print("!!! Automatic sampler selection is work-in-progress. !!!")
```
Comment (amal-ghamdi) [inline cuqi/problem/_problem.py:320]:
```suggestion
        print("!!! Automatic sampler selection is a work-in-progress. !!!")
```

**Review by amal-ghamdi — CHANGES_REQUESTED** [review]
Thanks @nabriis! I add a few suggested changes and added issue #505 to capture the discussion here.

File: cuqi/problem/_problem.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (amal-ghamdi) [reply cuqi/problem/_problem.py:896]: Sounds good, I added issue #505 to capture this

**Review by amal-ghamdi — APPROVED** [review]
LGTM!

**Pull request merged**